crypto/dlogproofs/partial_dlog_knowledge_ec.go

```diff
@@ -0,0 +1,160 @@
+	z2               *big.Int
+	ord				 int 
+}
+
+func NewPartialECDLogProver(dlog *dlog.ECDLog) *PartialECDLogProver {
+	var prover PartialECDLogProver
+	prover = PartialECDLogProver{
+		DLog:         dlog,
+	}
+
+	return &prover
+}
+
+func (prover *PartialECDLogProver) GetProofRandomData(secret1 *big.Int, a1, b1, a2, 
+		b2 *types.ECGroupElement) (*types.ECGroupElement, *types.ECGroupElement,
+		*types.ECGroupElement, *types.ECGroupElement,
+		*types.ECGroupElement, *types.ECGroupElement) {
+	prover.a1 = a1
+	prover.a2 = a2
+	prover.secret1 = secret1
+	r1 := common.GetRandomInt(prover.DLog.GetOrderOfSubgroup())
+	c2 := common.GetRandomInt(prover.DLog.GetOrderOfSubgroup())
+	z2 := common.GetRandomInt(prover.DLog.GetOrderOfSubgroup())
+	prover.r1 = r1
+	prover.c2 = c2
+	prover.z2 = z2
+	x1X, x1Y := prover.DLog.Exponentiate(a1.X, a1.Y, r1)
+	x2X, x2Y := prover.DLog.Exponentiate(a2.X, a2.Y, z2)
+	b2ToC2X, b2ToC2Y := prover.DLog.Exponentiate(b2.X, b2.Y, c2)
+	b2ToC2InvX, b2ToC2InvY := prover.DLog.Inverse(b2ToC2X, b2ToC2Y)
+	x2X, x2Y = prover.DLog.Multiply(x2X, x2Y, b2ToC2InvX, b2ToC2InvY)
+	
+	x1 := types.NewECGroupElement(x1X, x1Y)
+	x2 := types.NewECGroupElement(x2X, x2Y)
+	
+	// we need to make sure that the order does not reveal which secret we do know:
+	ord := common.GetRandomInt(big.NewInt(2))
+	
+	if ord.Cmp(big.NewInt(0)) == 0 {
+		prover.ord = 0
+		return x1, a1, b1, x2, a2, b2
+	} else {
+		prover.ord = 1
+		return x2, a2, b2, x1, a1, b1
+	}
+}
+
+func (prover *PartialECDLogProver) GetProofData(challenge *big.Int) (*big.Int, *big.Int,
+		*big.Int, *big.Int) {
+	c1 := new(big.Int).Xor(prover.c2, challenge)	
+
+	z1 := new(big.Int)
+	z1.Mul(c1, prover.secret1)
+	z1.Add(z1, prover.r1)
+	z1.Mod(z1, prover.DLog.GetOrderOfSubgroup())
+	
+	if prover.ord == 0 {
+		return c1, z1, prover.c2, prover.z2
+	} else {
+		return prover.c2, prover.z2, c1, z1
+	}
+}
+
+type PartialECDLogVerifier struct {
+	DLog              *dlog.ECDLog
+	x1                *types.ECGroupElement
+	x2                *types.ECGroupElement
+	a1                *types.ECGroupElement
+	b1                *types.ECGroupElement
+	a2                *types.ECGroupElement
+	b2                *types.ECGroupElement
+	challenge         *big.Int
+}
+
+func NewPartialECDLogVerifier(dlog *dlog.ECDLog) *PartialECDLogVerifier {
+	verifier := PartialECDLogVerifier{
+		DLog:         dlog,
+	}
+
+	return &verifier
+}
+
+func (verifier *PartialECDLogVerifier) SetProofRandomData(x1, a1, b1, x2, a2, 
+		b2 *types.ECGroupElement) {
+	verifier.x1 = x1
+	verifier.a1 = a1
+	verifier.b1 = b1
+	verifier.x2 = x2
+	verifier.a2 = a2
+	verifier.b2 = b2
+}
+
+func (verifier *PartialECDLogVerifier) GetChallenge() *big.Int {
+	challenge := common.GetRandomInt(verifier.DLog.GetOrderOfSubgroup())
+	verifier.challenge = challenge
+	return challenge
+}
+
+func (verifier *PartialECDLogVerifier) Verify(c1, z1, c2, z2 *big.Int) bool {
+	c := new(big.Int).Xor(c1, c2)	
+	if c.Cmp(verifier.challenge) == 0 {
```

Comment:
What do you mean exactly? It needs to be changed whether c = c1 xor c2 (c is chosen by verifier, while c1 and c2 by prover).

Reply:
Sorry, I forgot to add a clarification. So, if we look at this block of code:
```go
if c.Cmp(verifier.challenge) == 0 {
} else {
  	return false	
}
```
if basically says "if the condition evaluates to true, do nothing. Else, return false". Then, if the condition previously evaluated to true, the code continues after the if-else statement.

I wanted to suggest that we change the if-else statement like this: 
```go
if c.Cmp(verifier.challenge) != 0 {
  	return false	
}
```
since if I am not mistaken, it does what you initially wanted to do with the if-else statement, right? If I am mistaken, please ignore this comment :slightly_smiling_face: . If you decide to change it, it occurs twice (here and in `crypto/dlogproofs/partial_dlog_knowledge.go` as well).

Reply:
Agree :)